### PR TITLE
Serializers

### PR DESCRIPTION
**Pull request by bbarberBPL**
Serializer module complete.

Adds 

- Inheritable Serializer Class
 
- DSL at the class level for per adapter schemas
 
- Ability to define/register custom schemas

- Conditionaaly add attribute `if/unless`

- Ability to whitelist included fields via `serializer_params`

- Ability to whitelist included relationships via `serializer params`

**TODO** 
 - Integrate this with serializers in the app.
- Write documentation/ create wiki page on serializer DSL

### CODE REVIEW

File: app/serializers/curator/institution_serializer.rb [hunk 1]
```diff
@@ -2,14 +2,16 @@
...
-    attributes :ark_id, :abstract, :name, :metastreams
-
-    def metastreams
-      {}.merge(
-        ActiveModelSerializers::SerializableResource.new(object.administrative, serializer: Metastreams::AdministrativeSerializer, root: 'administrative').as_json
-      ).merge(
-        ActiveModelSerializers::SerializableResource.new(object.workflow, serializer: Metastreams::WorkflowSerializer, root: 'workflow').as_json
-      )
+    schema_as_json do
+      attributes :ark_id, :abstract, :name, :metastreams
```
Comment (ebenenglish) [inline app/serializers/curator/institution_serializer.rb]: `metastreams` needs to be defined, or I get this error:
```
[21] pry(main)> Curator::InstitutionSerializer.new(inst, :json).render
NoMethodError: undefined method `metastreams' for #<Curator::Institution:0x000055e6ac89bca8>

```

File: app/serializers/curator/collection_serializer.rb [hunk 2]
```diff
@@ -2,14 +2,16 @@
...
-    attributes :ark_id, :abstract, :name, :metastreams
-
-    def metastreams
-      {}.merge(
-        ActiveModelSerializers::SerializableResource.new(object.administrative, serializer: Metastreams::AdministrativeSerializer, root: 'administrative').as_json
-      ).merge(
-        ActiveModelSerializers::SerializableResource.new(object.workflow, serializer: Metastreams::WorkflowSerializer, root: 'workflow').as_json
-      )
+    schema_as_json root: :collection do
+      attributes :ark_id, :abstract, :name, :metastreams
```
Comment (ebenenglish) [inline app/serializers/curator/collection_serializer.rb]: `metastreams` needs to be defined, or we get an error:
```
Curator::CollectionSerializer.new(col, :json).render
NoMethodError: undefined method `metastreams' for #<Curator::Collection:0x000055e6aa265f98>
```

File: spec/lib/curator/serializers/adapters/json_adapter_spec.rb [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        it 'is expected to be parsed as json' do
+          expect(Oj.load(subject)).to be_a_kind_of(ActiveSupport::HashWithIndifferentAccess)
+          expect(Oj.load(subject)).to match(a_hash_including(record_hash.call(institution).as_json.with_indifferent_access))
+        end
+      end
+
+      describe 'collection' do
+        subject { json_adapter_no_root.render(institutions) }
+
+        let(:records_array) { institutions.map { |inst| record_hash.call(inst).as_json.with_indifferent_access } }
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply spec/lib/curator/serializers/adapters/json_adapter_spec.rb]: resolved in [ccdf6bf](https://github.com/boston-library/curator/pull/54/commits/ccdf6bf29dad0d8df83d0a7ed780fa80485b6c50)

File: spec/lib/curator/serializers/adapters/json_adapter_spec.rb [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        it 'is expected to be parsed as json' do
+          expect(Oj.load(subject)).to be_a_kind_of(ActiveSupport::HashWithIndifferentAccess).and have_key(:institution)
+          expect(Oj.load(subject)).to match('institution' => a_hash_including(record_hash.call(institution).as_json.with_indifferent_access))
+        end
+      end
+
+      describe 'collection' do
+        subject { json_adapter_with_root.render(institutions) }
+
+        let(:records_array) { institutions.map { |inst| record_hash.call(inst).as_json.with_indifferent_access } }
```
[the remark this answers is not in the document]
Comment (bbarberBPL) [reply spec/lib/curator/serializers/adapters/json_adapter_spec.rb]: resolved in [ccdf6bf](https://github.com/boston-library/curator/pull/54/commits/ccdf6bf29dad0d8df83d0a7ed780fa80485b6c50)

**Pull request merged**